internal/kubernetes/token/grpc.go

```diff
@@ -0,0 +1,151 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package token
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+	authv1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type tokenResolver struct {
+	kubeclient     client.Client
+	namespace      string
+	serviceAccount string
+
+	token      string
+	expiration metav1.Time
+}
+
+func WithServiceAccountToken(client client.Client, namespace, serviceAccount string) grpc.DialOption {
+	tr := tokenResolver{
+		kubeclient:     client,
+		namespace:      namespace,
+		serviceAccount: serviceAccount,
+		expiration:     metav1.Now(),
+	}
+
+	return grpc.WithUnaryInterceptor(tr.addAuthorizationHeader)
+}
+
+func (tr *tokenResolver) addAuthorizationHeader(ctx context.Context, method string, req, reply any, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+	token, err := tr.getToken(ctx)
+	if err != nil {
+		return err
+	}
+
+	authCtx := metadata.AppendToOutgoingContext(ctx, "Authorization", "Bearer "+token)
+	return invoker(authCtx, method, req, reply, cc, opts...)
+}
+
+func (tr *tokenResolver) getToken(ctx context.Context) (string, error) {
+	now := metav1.Now()
+	if tr.expiration.Before(&now) {
+		// token expired
+		return tr.refreshToken(ctx)
+	}
+
+	return tr.token, nil
+}
+
+func (tr *tokenResolver) refreshToken(ctx context.Context) (string, error) {
+	treq := &authv1.TokenRequest{
+		Spec: authv1.TokenRequestSpec{
+			Audiences: []string{"csi-addons"},
+		},
+	}
+
+	err := tr.kubeclient.Create(ctx, treq)
+	if err != nil {
+		return "", err
+	}
+
+	tr.token = treq.Status.Token
+	tr.expiration = treq.Status.ExpirationTimestamp
+
+	return tr.token, nil
+}
+
+func AuthorizationInterceptor(kubeclient kubernetes.Clientset) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		if err := authorizeConnection(kubeclient, ctx); err != nil {
+			return nil, err
+		}
+		return handler(ctx, req)
+	}
+}
+
+func authorizeConnection(kubeclient kubernetes.Clientset, ctx context.Context) error {
+
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return status.Errorf(codes.Unauthenticated, "missing metadata")
+	}
+
+	authHeader, ok := md["authorization"]
+	if !ok || len(authHeader) == 0 {
+		return status.Errorf(codes.Unauthenticated, "missing authorization token")
+	}
+
+	token := authHeader[0]
+	isValidated, err := validateBearerToken(token, kubeclient)
```

Comment:
pass the `ctx` to `validateBearerToken` so that there is no `context.TODO()` needed there?